core/src/main/java/org/elasticsearch/tasks/Task.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.action.admin.cluster.node.tasks.list.TaskInfo;
+import org.elasticsearch.common.inject.Provider;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Current task information
+ */
+public class Task {
+
+    private final long id;
+
+    private final String type;
+
+    private final String action;
+
+    private final Provider<String> description;
+
+    private final AtomicBoolean canceled;
+
+    private Listener listener;
+
+    public Task(long id, String type, String action, Provider<String> description) {
+        this.id = id;
+        this.type = type;
+        this.action = action;
+        this.description = description;
+        canceled = new AtomicBoolean();
+    }
+
+    /**
+     * Returns task id
+     */
+    public long id() {
+        return id;
+    }
+
+    /**
+     * Returns task channel type (netty, transport, direct)
+     */
+    public String type() {
+        return type;
+    }
+
+    /**
+     * Returns task action
+     */
+    public String action() {
+        return action;
+    }
+
+    /**
+     * Generates task description
+     */
+    public String description() {
+        return description.get();
+    }
+
+    /**
+     * Marks task as cancelled
+     */
+    public void cancel() {
+        if (canceled.getAndSet(true) == false && listener != null) {
```

Comment:
This is left over from the cancellation part of PR that I removed in order to make this PR more manageable. I am going to clean up this cancellation-related pieces from this PR here for now and will open another PR with fully functional cancellation logic there.